### PR TITLE
new assertion algorithm for ngspice-watcher-test, closes #183

### LINKED ISSUE #183: Tests fail locally
### Steps to reproduce
1. git clone ... && git checkout b29e8dbf080b740bd49e85d5b9351dfc3a0737f2
2. ./waf configure build
3. ./waf

### Expected behaviour
Success, exit 0

### Actual behaviour
```
/home/bernhard/Projects/oregano/build/src/microtests                                                  10:00:03   1 
/core/coords: OK
/core/engine: OK
/core/model/nodestore: OK
/core/model/wire/intersection: OK
/core/model/wire/tcrossing: OK
/core/model/part-property/update_connection_designators/basic: OK
/core/model/part-property/update_connection_designators/AT: OK
/core/model/part-property/update_connection_designators/AMPERSAND_TRUE: OK
/core/model/part-property/update_connection_designators/AMPERSAND_FALSE: OK
/core/model/part-property/update_connection_designators/QUESTION_MARK_TRUE: OK
/core/model/part-property/update_connection_designators/QUESTION_MARK_FALSE: OK
/core/model/part-property/update_connection_designators/TILDE_TRUE: OK
/core/model/part-property/update_connection_designators/TILDE_FALSE: OK
/core/model/part-property/update_connection_designators/HASHTAG_TRUE: OK
/core/model/part-property/update_connection_designators/HASHTAG_FALSE: OK
/core/engine/ngspice/watcher/basic: ../src/engines/ngspice-analysis.c:832 @ ngspice_analysis +++ 0 buf = Transient Analysis  Sat Jun 17 10:00:12  2017
**
ERROR:../test/test_engine_ngspice.c:154:test_engine_ngspice_basic: 'distance < 3*16*20' should be TRUE
[1]    17043 abort (core dumped)  /home/bernhard/Projects/oregano/build/src/microtests


```

### Configuration
**Operating system**:
```
cat /etc/os-release |grep PRETTY_NAME
PRETTY_NAME="Fedora 25 (Workstation Edition)"
```

**Versions**
```
oregano --version
Oregano 0.84.5
 Website: https://ahoi.io/projects/oregano
 License: GPLv2/GPLv3
 Main Developer: Bernhard Schuster
```
```
ngspice --version

ngspice compiled from ngspice revision 26
Written originally by Berkeley University
Currently maintained by the NGSpice Project

Copyright (C) 1985-1996,  The Regents of the University of California
Copyright (C) 1999-2011,  The NGSpice Project
```

### PR DESCRIPTION
**Pull request by Michi1994**
# Todo

- [x] parse expected-file to RAM by using `void ngspice_analysis (NgspiceAnalysisResources *resources)`
- [x] implement function-metric
- [x] compare expected data with actual data (RAM)
- [x] compare actual data (RAM) with actual data (SSD/HDD)

### CODE REVIEW

**Comment by Michi1994** [comment]
Is there any library to handle X-Y-data in C? I want to

- resample
- scale in X direction
- scale in Y direction
- subtract
- integrate

two X-Y datas.

**Comment by drahnr** [comment]
Not at the top of my head, usually, if the problem is small and contained, you write it yourself, if it is perf relevant you end up with a huge dependency.

**Comment by Michi1994** [comment]
Why is the output of [https://ci.spearow.io/teams/main/pipelines/oregano/jobs/pr-compile/builds/10](https://ci.spearow.io/teams/main/pipelines/oregano/jobs/pr-compile/builds/10) hiding the failure output?

How can I access `build/test.log`?

**Comment by Michi1994** [comment]
It's working again: [https://ci.spearow.io/teams/main/pipelines/oregano/jobs/pr-compile/builds/11](https://ci.spearow.io/teams/main/pipelines/oregano/jobs/pr-compile/builds/11)

**Comment by Michi1994** [comment]
This pull request is ready for review.

**Comment by drahnr** [comment]
You need to rebase and get rid of all those ci test commits and squash all the commits with the same name.

File: .concourse/fedora/buildrpm.sh [hunk 1]
```diff
@@ -2,6 +2,11 @@
 
 set -x
 set -e
+set -x
+
+pwd 2>&1
+RPMBUILD_DIR="$(pwd)/../srpm/rpmbuild/"
+mkdir -p ${RPMBUILD_DIR}/{SOURCES,BUILD,RPMS,SRPMS,SPECS}
```
Comment (drahnr) [inline .concourse/fedora/buildrpm.sh]: code duplication

File: .gitignore [hunk 2]
```diff
@@ -11,3 +11,4 @@ waftools/__pycache__*
 .waf*
 credentials.yml
 .credentials.yml
+/.project
```
Comment (drahnr) [inline .gitignore]: plase use `.project` instead

File: src/engines/ngspice-watcher.c [hunk 3]
```diff
@@ -302,7 +302,7 @@ enum NGSPICE_WATCHER_RETURN_VALUE {
 static enum NGSPICE_WATCHER_RETURN_VALUE
 ngspice_watcher_watch_ngspice_resources (GPid pid, gint status, NgSpiceWatcherWatchNgSpiceResources *resources) {
 	GThread *worker = resources->worker;
-	GThread *saver = resources->saver;
+	GThread **saver = resources->saver;
```
Comment (drahnr) [inline src/engines/ngspice-watcher.c]: add a ref each, and unref, this makes sure the thread is always valid (even if the side conditions already might make sure that this is the case)

File: src/engines/ngspice.c [hunk 4]
```diff
@@ -94,17 +94,10 @@ static void ngspice_class_init (OreganoNgSpiceClass *klass)
...
-{
-	OreganoNgSpice *ngspice;
-	GList *iter;
-	int i;
-
-	ngspice = OREGANO_NGSPICE (object);
-	iter = ngspice->priv->analysis;
-	for (; iter; iter = iter->next) {
+void ngspice_analysis_finalize(GList *analysis) {
+	for (GList *iter = analysis; iter; iter = iter->next) {
```
Comment (drahnr) [inline src/engines/ngspice.c]: How did you compile this? I think you need C99+ to be able to do inline for loop declarations ey?

Comment (drahnr) [inline src/engines/ngspice.c]: We can make this a requirement, but I just want to make sure there is awareness of that.

File: test/test_engine_ngspice.c [hunk 5]
```diff
@@ -44,6 +46,11 @@ static void test_engine_ngspice_log_append_error(GList **list, const gchar *stri
 	*list = g_list_append(*list, g_strdup(string));
 }
 
+static gboolean test_engine_ngspice_timeout(GMainLoop *loop) {
+	g_main_loop_quit(loop);
+	return FALSE;
+}
```
Comment (drahnr) [inline test/test_engine_ngspice.c]: please use G_SOURCE_CONTINUE or G_SOURCE_REMOVE

File: test/test_engine_ngspice.c [hunk 6]
```diff
@@ -195,29 +351,28 @@ static void test_engine_ngspice_error_step_zero() {
...
-  // 	GList *walker = test_resources->log_list;
-  //
-  // 	for (int i = 0; array[i] != NULL; i++) {
-  // 		g_assert_nonnull(walker);
-  // 		g_assert_nonnull(walker->data);
-  // 		g_assert_cmpstr(walker->data, ==, array[i]);
-  // 		walker = walker->next;
-  // 	}
+	g_assert_cmpstr(actual_content, ==, expected_content);
+
```
Comment (drahnr) [inline test/test_engine_ngspice.c]: comparing for equality is no solution, I can imaginge that even arch specific rounding might cause this to fail

**Comment by drahnr** [comment]
I will adjust the command that if it fails, the generated `test.log` will be printed.

**Comment by drahnr** [comment]
Any progress or ETA?

File: test/test_engine_ngspice.c (hunk 6 above)
Comment (drahnr) [reply test/test_engine_ngspice.c:355]: `_cmpstr`, so it's fine.

**Comment by drahnr** [comment]
I apologize for the - in retrospective - overly harsh review comments. I'd be still happy to get this merged.

File: src/engines/ngspice-watcher.c (hunk 3 above)
Comment (drahnr) [reply src/engines/ngspice-watcher.c:305]: Not sure what I meant with that :shrug:

**Pull request closed without merging**